### PR TITLE
Move type related FAQ points

### LINKED ISSUE #1219: Documentation: merge the FAQ in the appropriate sections
There are a lot of pieces of information which would be better to be merged at the appropriate locations. A lot is already duplicated. Many of the examples refer to outdated features (e.g. `import "mortal";`).

### LINKED ISSUE #2178: doc: explain that state variables can be initialized inline
[Solidity features](https://github.com/ethereum/wiki/wiki/Solidity-Features) wiki page contained the following section.  The documentation in this repository does not explain this feature yet.  This issue keeps track of adding such explanations to the this repository.

> Inline members initialization
> 
> This point seems to be missing from the official documentation.
> 
> PT Inline members can be initialized at declaration time.
> 
> contract test {
>   function test(){
>     m_b = 6;
>   }
>   uint m_a = 5;
>   uint m_b;
> }

### PR DESCRIPTION
**Pull request by ChrisChinchilla**
Partially closes https://github.com/ethereum/solidity/issues/1219 including some formatting and language fixes in related points, additional questions below.

Also closes https://github.com/ethereum/solidity/issues/2178.

### CODE REVIEW

File: docs/types.rst [hunk 1]
```diff
@@ -609,6 +646,21 @@ number of bytes, always use one of ``bytes1`` to ``bytes32`` because they are mu
     that you are accessing the low-level bytes of the UTF-8 representation,
     and not the individual characters!
 
+You can initialise a statically sized memory array inline using syntax such as ``string[] memory myarray = new string[](4);``. For example.
```
Comment (ChrisChinchilla) [inline docs/types.rst]: This isn't quite complete, I'd like to match the text and the example.

File: docs/types.rst [hunk 2]
```diff
@@ -984,3 +1027,24 @@ converted to a matching size. This makes alignment and padding explicit::
     bytes32(uint256(x)); // pad on the left
     bytes32(bytes2(x)); // pad on the right
 
+.. index:: ! type;deduction, ! var
+
+.. _type-deduction:
+
+Type Deduction
```
Comment (ChrisChinchilla) [inline docs/types.rst]: Not sure where this came from as I didn't add it, possibly from develop branch.

File: docs/types.rst [hunk 3]
```diff
@@ -6,12 +6,11 @@
...
-Solidity is a statically typed language, which means that the type of each
-variable (state and local) needs to be specified.
-Solidity provides several elementary types which can be combined to form complex types.
+Solidity is a statically typed language, which means that you need to specify the type of each variable (state and local). Solidity provides elementary types which you can combine to form complex types.
 
-In addition, types can interact with each other in expressions containing
-operators. For a quick reference of the various operators, see :ref:`order`.
+Types can interact in expressions containing operators. For a quick reference of the various operators, see :ref:`order`.
+
+Types can not return an ``undefined`` or "null" value. To handle any unexpected values you can 'throw' on error, which also reverts the whole transaction.
```
Comment (chriseth) [inline docs/types.rst]: Better: The concept of "undefined" or "null" does not exist in Solidity.

Also: Don't mention throw, rather use `revert("Error message")`.

File: docs/types.rst [hunk 4]
```diff
@@ -52,6 +51,12 @@ Operators:
 * Bit operators: ``&``, ``|``, ``^`` (bitwise exclusive or), ``~`` (bitwise negation)
 * Arithmetic operators: ``+``, ``-``, unary ``-``, unary ``+``, ``*``, ``/``, ``%`` (remainder), ``**`` (exponentiation), ``<<`` (left shift), ``>>`` (right shift)
 
+The bit size of an integer determines its range, for example, for ``uint256``, this is 0 up to 2 :sup:`256` -1. If the result of an operation doesn't fit inside this range, it is truncated. These truncations can have `serious consequences <https://en.bitcoin.it/wiki/Value_overflow_incident>`_, so use code like the below to avoid certain attacks and check a value is what you expect it to be.
```
Comment (chriseth) [inline docs/types.rst]: The truncations themselves do not have serious consequences, rather not taking them into account might.

File: docs/types.rst [hunk 5]
```diff
@@ -52,6 +51,12 @@ Operators:
...
+
+::
+
+    require((balanceOf[_to] + _value) >= balanceOf[_to]);
```
Comment (chriseth) [inline docs/types.rst:58]: This only works for checking overflow for unsigned types, so we should make that clear.

File: docs/types.rst [hunk 6]
```diff
@@ -296,16 +301,26 @@ a non-rational number).
...
 ---------------
 
-String literals are written with either double or single-quotes (``"foo"`` or ``'bar'``).  They do not imply trailing zeroes as in C; ``"foo"`` represents three bytes not four.  As with integer literals, their type can vary, but they are implicitly convertible to ``bytes1``, ..., ``bytes32``, if they fit, to ``bytes`` and to ``string``.
+String literals are written with either double or single-quotes (``"foo"`` or ``'bar'``).  They do not imply trailing zeroes as in C; ``"foo"`` represents three bytes not four.  As with integer literals, their type can vary, but they are implicitly convertible to ``bytes1``, ..., ``bytes32``, if they fit, to ``bytes`` and to ``string``. For example, with ::
 
-String literals support escape characters, such as ``\n``, ``\xNN`` and ``\uNNNN``. ``\xNN`` takes a hex value and inserts the appropriate byte, while ``\uNNNN`` takes a Unicode codepoint and inserts an UTF-8 sequence.
+  bytes32 samevar = "stringliteral";
 
-.. index:: literal, bytes
+The string literal is interpreted in its raw byte form and if you inspect ``somevar``, you see a 32-byte hex value, which is the ``"stringliteral"`` in hex.
```
Comment (chriseth) [inline docs/types.rst]: Whether or not it is hex depends on the tool you use.

File: docs/types.rst [hunk 7]
```diff
@@ -296,16 +301,26 @@ a non-rational number).
...
+
+Using a ``string`` type is basically identical to ``bytes``, but it is assumed to hold the UTF-8 encoding of a real string. Since ``string`` UTF-8 encoding of some characters takes more than a single byte, it is quite expensive to compute the number of characters in the string. Because of this, Solidity does not support length methods ``string s; s.length``, or accessing an item at an array index ``s[2]``. If you want to access the low-level byte encoding of the string, you can use ``bytes(s).length`` and ``bytes(s)[2]`` which returns the number of bytes in the UTF-8 encoding of the string (not the number of characters) and the second byte (not character) of the UTF-8 encoded string, respectively.
```
Comment (chriseth) [inline docs/types.rst]: Not only "using" but also the type itself is basically identical to `bytes`.

Note that `strings` is not a string literal, it is a string type, so we should probably move this somewhere else.

File: docs/types.rst [hunk 8]
```diff
@@ -296,16 +301,26 @@ a non-rational number).
...
+
+  bytes32 samevar = "stringliteral";
+
+The string literal is interpreted in its raw byte form and if you inspect ``somevar``, you see a 32-byte hex value, which is the ``"stringliteral"`` in hex.
+
+Using a ``string`` type is basically identical to ``bytes``, but it is assumed to hold the UTF-8 encoding of a real string. Since ``string`` UTF-8 encoding of some characters takes more than a single byte, it is quite expensive to compute the number of characters in the string. Because of this, Solidity does not support length methods ``string s; s.length``, or accessing an item at an array index ``s[2]``. If you want to access the low-level byte encoding of the string, you can use ``bytes(s).length`` and ``bytes(s)[2]`` which returns the number of bytes in the UTF-8 encoding of the string (not the number of characters) and the second byte (not character) of the UTF-8 encoded string, respectively.
+
+String literals support escape characters, such as ``\n``, ``\xNN`` and ``\uNNNN``. ``\xNN`` takes a hex value and inserts the appropriate byte, while ``\uNNNN`` takes a Unicode codepoint and inserts an UTF-8 sequence.
```
Comment (chriseth) [inline docs/types.rst]: Pleas clarify whether `\u` takes hex or decimal.

File: docs/types.rst [hunk 9]
```diff
@@ -575,17 +589,40 @@ Default data location:
  - parameters (also return) of functions: memory
  - all other local variables: storage
 
+
+.. index:: copy an array, copy a struct
+
+Copying Reference Types
```
Comment (chriseth) [inline docs/types.rst]: I think this should also go into the newly created "copying, references, function calls, whatever" section with a link from here.

File: docs/types.rst (hunk 1 above)
Comment (chriseth) [reply docs/types.rst]: Perhaps better to move this into "Array literals / inline arrays"

File: docs/types.rst (hunk 8 above)
Comment (ChrisChinchilla) [reply docs/types.rst]: I think done. I couldn't find any definitive Solidity doc mentioning it, so extrapolated from C and JavaScript docs.

File: docs/types.rst (hunk 7 above)
Comment (ChrisChinchilla) [reply docs/types.rst]: Attempted to split and consolidate, more changes might be required, again, probably in a new PR.

File: docs/types.rst (hunk 6 above)
Comment (ChrisChinchilla) [reply docs/types.rst]: Clarified, might be nice to add a list at some point, will save that for a separate issue.

File: docs/types.rst (hunk 5 above)
Comment (ChrisChinchilla) [reply docs/types.rst:58]: Done

File: docs/types.rst (hunk 4 above)
Comment (ChrisChinchilla) [reply docs/types.rst]: Done

File: docs/types.rst (hunk 3 above)
Comment (ChrisChinchilla) [reply docs/types.rst]: Done

File: docs/types.rst (hunk 1 above)
Comment (ChrisChinchilla) [reply docs/types.rst]: Done

File: docs/types.rst (hunk 9 above)
Comment (ChrisChinchilla) [reply docs/types.rst]: @chriseth I think we should make this a new PR, so I'll revert this addition, keep the removal from the FAQ document and I'll stash this text for a forthcoming new PR.

File: docs/types.rst [hunk 10]
```diff
@@ -52,6 +51,12 @@ Operators:
 * Bit operators: ``&``, ``|``, ``^`` (bitwise exclusive or), ``~`` (bitwise negation)
 * Arithmetic operators: ``+``, ``-``, unary ``-``, unary ``+``, ``*``, ``/``, ``%`` (remainder), ``**`` (exponentiation), ``<<`` (left shift), ``>>`` (right shift)
 
+The bit size of an unsigned integer determines its range, for example, for ``uint256``, this is 0 up to 2 :sup:`256` -1. If the result of an operation doesn't fit inside this range, it is truncated. Not taking these truncations into account can have `serious consequences <https://en.bitcoin.it/wiki/Value_overflow_incident>`_, so use code like the below to avoid attacks and check a value is what you expect it to be.
+
+::
+
+    require((balanceOf[_to] + _value) >= balanceOf[_to]);
```
Comment (axic) [inline docs/types.rst:58]: I think this is a bad idea without explanation.

Comment (ChrisChinchilla) [reply docs/types.rst:58]: @axic I thought the text above and the example were enough, but added a little more, hopefully I got it right.

**Pull request closed without merging**